### PR TITLE
New canonical tag helper

### PR DESCRIPTION
**Pull request by 1v9**

### CODE REVIEW

**Comment by stevenjoezhang** [comment]
Wait for https://github.com/hexojs/hexo/pull/3691

**Comment by 1v9** [comment]
https://github.com/hexojs/hexo-pagination/blob/8b91a5cc0ddb0dfd0cbd1d708322569803eb23ef/lib/pagination.js#L16

https://github.com/hexojs/hexo-generator-index/blob/569bbb2c22a4d595682ba1014f5192173b4b7e36/lib/generator.js#L13

**Comment by 1v9** [comment]
`{{ page.permalink }}` works on every page exclude index and archive page.
`{{ url }}` is everythere.

**Comment by curbengh** [comment]
> Wait for [hexojs/hexo#3691](https://github.com/hexojs/hexo/pull/3691)

Most likely the option won't be backported to Hexo v3. Besides, this PR doesn't conflict with that option; assuming `index.html` only appear once at the end of url, `String.replace(/index\.html$/, '').replace(/index\.html$/, '')` still produces similar result.

File: scripts/helpers/engine.js [hunk 1]
```diff
@@ -4,17 +4,31 @@
...
-
 hexo.extend.helper.register('hexo_env', type => hexo.env[type]);
 
 hexo.extend.helper.register('next_env', type => {
   const path = require('path');
   const env = require(path.normalize('../../package.json'));
   return env[type];
 });
+
+hexo.extend.helper.register('canonical', function() {
```
Comment (1v9) [inline scripts/helpers/engine.js]: hexo.extend.helper.register('canonical', function() {
  const permalink = hexo.config.permalink;
  const canonical = hexo.theme.config.canonical;
  if (!canonical) return '';
  var url = this.url.replace('index.html', '');
  if (permalink.endsWith('.html')) {
    return `<link rel="canonical" href="${url}">`;
  }
  return `<link rel="canonical" href="${url.replace('.html', '')}">`;
});

**Comment by stevenjoezhang** [comment]
https://github.com/hexojs/hexo/pull/3691#issuecomment-532677671

permalink (endsWith '.html')
trailing_html
trailing_slash

What's the difference? Which is the recommended one?

**Comment by stevenjoezhang** [comment]
@SukkaW Thank you. It'll be great if these features can be added to Hexo at the next release

**Comment by stevenjoezhang** [comment]
Supersedes by https://github.com/theme-next/hexo-theme-next/pull/1175

**Pull request merged**